File: Non-procedural (functional) programming/Solved/antihric/male/dfs_casy.hs
module DfsCasy(dfsCasy) where
import Functional

--http://forum.matfyz.info/viewtopic.php?f=169&t=11756&p=41531&hilit=dfs#p41531 (akorat zadano v prologu)

data Graf a = G [a] [(a,a)] deriving Show
testGraf = G ['a','b','c','d','e','s'] [('a','s'),('a','b'), ('a','c'), ('b','d'), ('b','e'), ('c','e'), ('d','e')]

sousedi :: Eq a => Graf a -> a -> [a]
sousedi (G v e) u = foldr (\(x,y) a -> if x == u then y:a else if y == u then x:a else a) [] e

dfsCasy :: Eq a => Graf a -> a -> [(a, Int, Int)]
dfsCasy g a = v
        where
            (_,_, v) = dfsCasy' g [a] 0

dfsCasy' :: Eq a => Graf a -> [a] -> Int -> ([a] ,Int, [(a, Int, Int)])
dfsCasy' g xs@(x : _) c | null pripustniSousedi = (xs,c+2, [(x,c,c+1)])
                        | otherwise             = (nvisited,cislo+1, (x,c,cislo):vs)                     
        where
            d = c + 1
            pripustniSousedi = filter (\y -> not $ elem y xs) (sousedi g x)
            (nvisited, cislo, vs) = zavolej pripustniSousedi xs g d []


zavolej :: Eq a => [a] -> [a] -> Graf a -> Int -> [(a, Int, Int)] -> ([a], Int, [(a, Int, Int)])
zavolej [] visited g d vd = (visited, d, vd)
zavolej (p : ps) visited g d vd  | elem p visited  = zavolej ps visited g d vd
                                 | otherwise       = zavolej ps nvisited g e (vd++vs)
        where (nvisited, e, vs) = dfsCasy' g (p:visited) d

File: Non-procedural Programming/Solved/antidvorak/male/dfs_casy.pl
?- consult('../functional.pl').

vstup(graf([a,b,c,d,e,s], [h(a,s), h(a,b), h(a,c), h(b,d), h(b,e),h(c,e),h(d,e)] )).

%tohle je hroznej humus
%http://forum.matfyz.info/viewtopic.php?f=169&t=11756&p=41531&hilit=dfs#p41531

obsahuje_souseda(U,h(U,B),Acc,[B|Acc]) :- !.
obsahuje_souseda(U,h(A,U),Acc,[A|Acc]) :- !.
obsahuje_souseda(_,_,Acc,Acc).

sousedi(G, U, Vs) :-   
    arg(2,G, E),
    foldl(obsahuje_souseda(U),E,[], Vs).

dfs_casy(G, V, Z) :- dfs_casy1(G, [V], 0, _-_-Z).


nenavstiven(Visited, V) :- \+ member(V, Visited).
                               %navstivene-dalsi_pouzitelne_cislo-ocislovane_vrcholy
dfs_casy1(G, [V | Visited], C, [V|Visited]-E-[V-C-D]) :-
    D is C+1,
    E is C+2,
    sousedi(G, V, Sousedi),
    filter(nenavstiven(Visited), Sousedi, PripustniS),
    PripustniS = [], !.

dfs_casy1(G, [V | Visited], C, NV-F-[V-C-NC|Vs]) :-
    D is C+1,
    sousedi(G, V, Sousedi),
    filter(nenavstiven(Visited), Sousedi, PripustniS),
    zavolej(PripustniS, [V|Visited], G, D, [], NV-NC-Vs),
    F is NC+1.    

zavolej([], Visited, _, D, Vd, Visited-D-Vd).
zavolej([P|Ps], Visited, G, D, Vd, Z) :-
    member(P,Visited), !,
    zavolej(Ps,Visited,G,D,Vd, Z).

zavolej([P|Ps], Visited, G, D, Vd, Z) :-
    dfs_casy1(G, [P|Visited], D, NV-E-Vs),
    append(Vd, Vs, V),
    zavolej(Ps, NV, G, E, V, Z).